test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,614 @@
+  t.plan(3)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function (done) {
+      t.ok('called')
+      done()
+    })
+    done()
+  }))
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function (done) {
+      t.ok('called')
+      done()
+    })
+
+    instance.addHook('onListen', function (done) {
+      t.ok('called')
+      done()
+    })
+
+    done()
+  }))
+
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+t.test('localhost onListen encapsulation should be called in order', t => {
+  t.plan(6)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, 'called in root')
+    t.equal(this.pluginName, fastify.pluginName, 'the this binding is the right instance')
+    done()
+  })
+
+  fastify.register(async (childOne, o) => {
+    childOne.addHook('onListen', function (done) {
+      t.equal(order++, 1, 'called in childOne')
+      t.equal(this.pluginName, childOne.pluginName, 'the this binding is the right instance')
+      done()
+    })
+    childOne.register(async (childTwo, o) => {
+      childTwo.addHook('onListen', async function () {
+        t.equal(order++, 2, 'called in childTwo')
+        t.equal(this.pluginName, childTwo.pluginName, 'the this binding is the right instance')
+      })
+    })
+  })
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+t.test('nonlocalhost onListen should be called in order', t => {
+  t.plan(2)
+
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 1, '2nd called in root')
+    done()
+  })
+  fastify.listen({
+    host: '::1',
+    port: 0
+  })
+})
+t.test('nonlocalhost async onListen should be called in order', async t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st async called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 1, '2nd async called in root')
+  })
+
+  await fastify.listen({
+    host: '::1',
+    port: 0
+  })
+})
+t.test('nonlocalhost onListen should manage error in sync', t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st sync called in root')
+    done(new Error('FAIL ON LISTEN'))
```

Comment:
we pass an error, and then?

i assume, if we throw the error like we discussed already, then the fastify.listen will become an unhandled Promise rejection.